File: legacy/dropzone47/bot.py
import asyncio
import os
from datetime import datetime, timezone
from typing import Any, Dict, List

from telegram import InlineKeyboardButton, InlineKeyboardMarkup, InputFile, Message, Update
from telegram.ext import (
    ApplicationBuilder,
    CallbackQueryHandler,
    CommandHandler,
    ContextTypes,
    MessageHandler,
    filters,
)
from yt_dlp import YoutubeDL

from .config import (
    AUDIO_KBITRATE,
    CLEANUP_AFTER_SEND,
    DOWNLOAD_DIR,
    MAX_CONCURRENT_DOWNLOADS,
    MAX_HEIGHT,
    RATE_LIMIT_MAX,
    RATE_LIMIT_WINDOW,
    TELEGRAM_MAX_MB,
    TELEGRAM_TOKEN,
    logger,
)
from .download import (
    find_output_files,
    force_remove,
    pick_files_for_choice,
    safe_cleanup,
    video_height_ladder,
    ytdlp_download_with_progress,
)
from .i18n import t
from .ratelimit import RateLimiter
from .session import delete_session, load_session, save_session, user_downloads, user_sessions
from .utils import (
    ensure_download_dir,
    has_enough_space,
    humanize_duration,
    is_valid_url,
    user_download_dir,
)

# Global concurrency cap: at most MAX_CONCURRENT_DOWNLOADS downloads run at once; the
# rest wait for a slot (an explicit queue). Created lazily to bind to the running loop.
_download_slots: "asyncio.Semaphore | None" = None
_rate_limiter = RateLimiter(RATE_LIMIT_MAX, RATE_LIMIT_WINDOW)


def _slots() -> asyncio.Semaphore:
    global _download_slots
    if _download_slots is None:
        _download_slots = asyncio.Semaphore(MAX_CONCURRENT_DOWNLOADS)
    return _download_slots


def _now() -> str:
    return datetime.now(timezone.utc).isoformat(timespec="seconds")


def _exceeds_size_limit(files: List[str]) -> bool:
    return any((os.path.getsize(fp) / (1024 * 1024)) > TELEGRAM_MAX_MB for fp in files)


async def send_files(
    chat_id: int, context: ContextTypes.DEFAULT_TYPE, title: str, files: List[str]
) -> None:
    for path in files:
        try:
            if path.lower().endswith(".mp3"):
                with open(path, "rb") as f:
                    await context.bot.send_audio(
                        chat_id=chat_id,
                        audio=InputFile(f, filename=os.path.basename(path)),
                        title=title,
                    )
            elif path.lower().endswith((".mp4", ".mkv", ".webm", ".mov")):
                with open(path, "rb") as f:
                    await context.bot.send_video(
                        chat_id=chat_id,
                        video=InputFile(f, filename=os.path.basename(path)),
                        supports_streaming=True,
                    )
            else:
                with open(path, "rb") as f:
                    await context.bot.send_document(
                        chat_id=chat_id,
                        document=InputFile(f, filename=os.path.basename(path)),
                    )
        except Exception as e:
            logger.warning("Failed to send %s: %s", path, e)
            await context.bot.send_message(
                chat_id=chat_id, text=t("send_failed", name=os.path.basename(path), error=e)
            )


async def start(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    message = update.effective_message
    if message is None:
        return
    await message.reply_text(t("welcome"))


async def handle_url(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    message = update.effective_message
    if message is None or message.text is None:
        return
    url = message.text.strip()
    user_id = message.from_user.id if message.from_user else 0

    if not is_valid_url(url):
        await message.reply_text(t("invalid_url"))
        return

    ensure_download_dir()
    await message.reply_text(t("fetching_info"))

    try:
        # Use a cache directory under DOWNLOAD_DIR to avoid $HOME perms
        cache = os.path.join(DOWNLOAD_DIR, ".cache", "yt-dlp")
        os.makedirs(cache, exist_ok=True)
        with YoutubeDL({"quiet": True, "cachedir": cache, "noplaylist": True}) as ydl:
            info = ydl.extract_info(url, download=False)
    except Exception as e:
        logger.warning("Info fetch failed for %s: %s", url, e)
        await message.reply_text(t("info_failed"))
        return

    if info is None:
        await message.reply_text(t("info_failed"))
        return

    title = info.get("title") or ""
    duration = info.get("duration")
    thumbnail = info.get("thumbnail")
    video_id = info.get("id") or ""

    # Store only what later steps need; the full yt-dlp info dict is large and would
    # bloat the in-memory cache and the shelve store.
    user_sessions[user_id] = {"url": url, "title": title, "id": video_id}
    save_session(user_id, user_sessions[user_id])

    keyboard = [
        [InlineKeyboardButton(t("btn_audio"), callback_data="audio")],
        [InlineKeyboardButton(t("btn_video"), callback_data="video")],
    ]
    reply_markup = InlineKeyboardMarkup(keyboard)

    caption = t("choose", title=title, duration=humanize_duration(duration))
    if thumbnail:
        await message.reply_photo(photo=thumbnail, caption=caption, reply_markup=reply_markup)
    else:
        await message.reply_text(text=caption, reply_markup=reply_markup)


async def download_and_send_task(
    user_id: int,
    chat_id: int,
    message_id: int,
    context: ContextTypes.DEFAULT_TYPE,
    choice: str,
    session: Dict[str, Any],
) -> None:
    title = session.get("title") or ""
    url = session.get("url") or ""
    vid_id = session.get("id") or ""
    dest_dir = user_download_dir(user_id)
    task = user_downloads.setdefault(
        user_id,
        {
            "status": "queued",
            "choice": choice,
            "title": title,
            "url": url,
            "id": vid_id,
            "files": [],
            "process": None,
            "cancel": False,
            "created_at": _now(),
            "updated_at": _now(),
        },
    )

    async def edit_caption(text: str) -> None:
        try:
            await context.bot.edit_message_caption(
                chat_id=chat_id, message_id=message_id, caption=text
            )
        except Exception:
            try:
                await context.bot.send_message(chat_id=chat_id, text=text)
            except Exception:
                pass

    try:
        # Wait for a global download slot (this is the queue). While waiting the task
        # stays "queued"; the user already saw the queued message.
        async with _slots():
            if task.get("cancel"):
                raise asyncio.CancelledError()
            task["status"] = "downloading"
            task["updated_at"] = _now()
            await edit_caption(t("downloading", title=title, choice=choice))

            min_free_mb = max(TELEGRAM_MAX_MB * 2, 2000)
            if not has_enough_space(min_free_mb):
                raise RuntimeError("Insufficient disk space for a safe download")

            if choice == "video":
                # Step down through the resolution ladder until the file fits the size
                # limit (or we run out of rungs, in which case we send best-effort).
                ladder = video_height_ladder(MAX_HEIGHT)
                files: List[str] = []
                for i, height in enumerate(ladder):
                    if i == 0:
                        label = "video"
                    else:
                        label = f"video ({height}p)"
                        await context.bot.send_message(
                            chat_id=chat_id, text=t("video_too_large", height=height)
                        )
                    await ytdlp_download_with_progress(
                        url,
                        "video",
                        max_height=height,
                        edit_caption_coro=edit_caption,
                        task=task,
                        label=label,
                        dest_dir=dest_dir,
                    )
                    files = pick_files_for_choice(find_output_files(vid_id, dest_dir), "video")
                    if not files:
                        raise RuntimeError("Downloaded video file not found")
                    if not _exceeds_size_limit(files):
                        break
                    if i < len(ladder) - 1:
                        force_remove(files)  # discard before retrying at a lower resolution
                task.setdefault("files", []).extend(files)
                task["status"] = "sending"
                task["updated_at"] = _now()
                await send_files(chat_id, context, title, files)

            if choice == "audio":
                await ytdlp_download_with_progress(
                    url,
                    "audio",
                    max_height=MAX_HEIGHT,
                    edit_caption_coro=edit_caption,
                    task=task,
                    label="audio",
                    dest_dir=dest_dir,
                )
                files = pick_files_for_choice(find_output_files(vid_id, dest_dir), "audio")
                if not files:
                    raise RuntimeError("Downloaded audio file not found")
                if _exceeds_size_limit(files):
                    await context.bot.send_message(
                        chat_id=chat_id, text=t("audio_too_large")
                    )
                    force_remove(files)  # discard before retrying at a lower bitrate
                    kbps = min(AUDIO_KBITRATE, 96)
                    await ytdlp_download_with_progress(
                        url,
                        "audio",
                        max_height=MAX_HEIGHT,
                        edit_caption_coro=edit_caption,
                        task=task,
                        label=f"audio ({kbps}kbps)",
                        dest_dir=dest_dir,
                        audio_kbps=kbps,
                    )
                    files = pick_files_for_choice(find_output_files(vid_id, dest_dir), "audio")
                    if not files:
                        raise RuntimeError("Audio file (lower bitrate) not found")
                task.setdefault("files", []).extend(files)
                task["status"] = "sending"
                task["updated_at"] = _now()
                await send_files(chat_id, context, title, files)

            task["status"] = "done"
            task["updated_at"] = _now()
            await edit_caption(t("completed", title=title))
    except asyncio.CancelledError:
        task["status"] = "canceled"
        task["updated_at"] = _now()
        await edit_caption(t("canceled"))
    except Exception as e:
        task["status"] = "error"
        task["updated_at"] = _now()
        await edit_caption(t("error", error=e))
    finally:
        if CLEANUP_AFTER_SEND and user_downloads.get(user_id, {}).get("files"):
            safe_cleanup(list(user_downloads[user_id]["files"]))
        delete_session(user_id)


async def handle_choice(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    query = update.callback_query
    if query is None:
        return
    await query.answer()
    user_id = query.from_user.id
    choice = query.data
    if choice not in {"audio", "video"}:
        return
    session = user_sessions.get(user_id) or load_session(user_id)

    if not session:
        await query.edit_message_text(t("session_not_found"))
        return

    # Old callbacks can carry an inaccessible message; bail if we can't act on it.
    if not isinstance(query.message, Message):
        return

    active = user_downloads.get(user_id)
    if active and active.get("status") in {"queued", "downloading", "sending"}:
        await query.message.reply_text(t("already_running"))
        return

    if not _rate_limiter.allow(user_id):
        minutes = max(1, _rate_limiter.retry_after(user_id) // 60)
        await query.message.reply_text(t("rate_limited", minutes=minutes))
        return

    chat_id = query.message.chat_id
    message_id = query.message.message_id
    task_coro = download_and_send_task(user_id, chat_id, message_id, context, choice, session)
    async_task = asyncio.create_task(task_coro)
    user_downloads[user_id] = {
        "async_task": async_task,
        "status": "queued",
        "choice": choice,
        "title": session.get("title"),
        "url": session.get("url"),
        "id": session.get("id"),
        "files": [],
        "process": None,
        "cancel": False,
        "created_at": _now(),
        "updated_at": _now(),
    }
    await query.edit_message_caption(
        caption=t("queued", title=session.get("title"), choice=choice)
    )


async def cmd_downloads(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    user_id = update.effective_user.id if update.effective_user else None
    if user_id is None:
        return
    message = update.effective_message
    if message is None:
        return
    task = user_downloads.get(user_id)
    if not task:
        await message.reply_text(t("no_downloads"))
        return
    await message.reply_text(
        t(
            "downloads_status",
            title=task.get("title"),
            choice=task.get("choice"),
            status=task.get("status"),
            created=task.get("created_at"),
            updated=task.get("updated_at"),
        )
    )


async def cmd_cancel(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    user_id = update.effective_user.id if update.effective_user else None
    if user_id is None:
        return
    message = update.effective_message
    if message is None:
        return
    task = user_downloads.get(user_id)
    if not task or task.get("status") not in {"queued", "downloading", "sending"}:
        await message.reply_text(t("nothing_to_cancel"))
        return
    # Cooperative cancellation: the yt-dlp progress hook observes this flag and aborts
    # the download thread cleanly. We deliberately do NOT call async_task.cancel(),
    # which would leave the executor thread running until the next hook fires anyway.
    task["cancel"] = True
    await message.reply_text(t("cancel_requested"))


async def cmd_clear_downloads(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    user_id = update.effective_user.id if update.effective_user else None
    if user_id is None:
        return
    message = update.effective_message
    if message is None:
        return
    task = user_downloads.get(user_id)
    if not task:
        await message.reply_text(t("nothing_to_clear"))
        return
    dest_dir = user_download_dir(user_id)
    ids: List[str] = [i for i in {task.get("id")} if i]
    removed = 0
    for vid in ids:
        for p in find_output_files(vid, dest_dir):
            try:
                os.remove(p)
                removed += 1
            except Exception as e:
                logger.warning("Failed to remove %s: %s", p, e)
    task["files"] = []
    await message.reply_text(t("cleared", removed=removed))


def run() -> None:
    app = ApplicationBuilder().token(TELEGRAM_TOKEN).build()
    app.add_handler(CommandHandler("start", start))
    app.add_handler(CommandHandler("downloads", cmd_downloads))
    app.add_handler(CommandHandler("cancel", cmd_cancel))
    app.add_handler(CommandHandler("clear_downloads", cmd_clear_downloads))
    app.add_handler(MessageHandler(filters.TEXT & ~filters.COMMAND, handle_url))
    app.add_handler(CallbackQueryHandler(handle_choice))
    app.run_polling()

File: internal/bot/bot.go
// Package bot wires the Telegram handlers, the concurrency queue, rate limiting,
// sessions and the download backends together.
package bot

import (
	"context"
	"fmt"
	"log/slog"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"

	tgbot "github.com/go-telegram/bot"
	"github.com/go-telegram/bot/models"

	"github.com/carlosprados/dropzone47/internal/config"
	"github.com/carlosprados/dropzone47/internal/download"
	"github.com/carlosprados/dropzone47/internal/i18n"
	"github.com/carlosprados/dropzone47/internal/ratelimit"
	"github.com/carlosprados/dropzone47/internal/session"
	"github.com/carlosprados/dropzone47/internal/util"
)

type task struct {
	status    string // queued | downloading | sending | done | error | canceled
	choice    string
	title     string
	id        string
	files     []string
	cancel    context.CancelFunc
	createdAt time.Time
	updatedAt time.Time
}

func (t *task) active() bool {
	return t.status == "queued" || t.status == "downloading" || t.status == "sending"
}

// Bot is the running Telegram bot.
type Bot struct {
	cfg     config.Config
	tr      *i18n.Translator
	dl      download.Downloader
	store   *session.Store
	limiter *ratelimit.Limiter
	slots   chan struct{}
	log     *slog.Logger

	mu    sync.Mutex
	tasks map[int64]*task
}

// New builds a Bot from its dependencies.
func New(cfg config.Config, dl download.Downloader, store *session.Store, log *slog.Logger) *Bot {
	return &Bot{
		cfg:     cfg,
		tr:      i18n.New(cfg.Lang),
		dl:      dl,
		store:   store,
		limiter: ratelimit.New(cfg.RateLimitMax, cfg.RateLimitWindow),
		slots:   make(chan struct{}, cfg.MaxConcurrentDownloads),
		log:     log,
		tasks:   make(map[int64]*task),
	}
}

// Run registers handlers and blocks on long polling until ctx is canceled.
func (b *Bot) Run(ctx context.Context) error {
	opts := []tgbot.Option{
		tgbot.WithDefaultHandler(b.onText),
		tgbot.WithCallbackQueryDataHandler("", tgbot.MatchTypePrefix, b.onCallback),
	}
	api, err := tgbot.New(b.cfg.TelegramToken, opts...)
	if err != nil {
		return err
	}
	api.RegisterHandler(tgbot.HandlerTypeMessageText, "/start", tgbot.MatchTypeExact, b.cmdStart)
	api.RegisterHandler(tgbot.HandlerTypeMessageText, "/downloads", tgbot.MatchTypeExact, b.cmdDownloads)
	api.RegisterHandler(tgbot.HandlerTypeMessageText, "/cancel", tgbot.MatchTypeExact, b.cmdCancel)
	api.RegisterHandler(tgbot.HandlerTypeMessageText, "/clear_downloads", tgbot.MatchTypeExact, b.cmdClear)

	b.log.Info("bot started", "downloader", b.dl.Name(), "lang", string(b.tr.Lang()))
	api.Start(ctx)
	return nil
}

func (b *Bot) cmdStart(ctx context.Context, api *tgbot.Bot, update *models.Update) {
	if update.Message == nil {
		return
	}
	b.send(ctx, api, update.Message.Chat.ID, b.tr.T("welcome"))
}

func (b *Bot) onText(ctx context.Context, api *tgbot.Bot, update *models.Update) {
	msg := update.Message
	if msg == nil || msg.From == nil {
		return
	}
	url := strings.TrimSpace(msg.Text)
	if strings.HasPrefix(url, "/") {
		return // unknown command
	}
	userID := msg.From.ID
	chatID := msg.Chat.ID

	if !util.IsValidURL(url) {
		b.send(ctx, api, chatID, b.tr.T("invalid_url"))
		return
	}
	b.send(ctx, api, chatID, b.tr.T("fetching_info"))

	infoCtx, cancel := context.WithTimeout(ctx, 60*time.Second)
	defer cancel()
	info, err := b.dl.FetchInfo(infoCtx, url)
	if err != nil || info.ID == "" {
		b.log.Warn("fetch info failed", "url", url, "err", err)
		b.send(ctx, api, chatID, b.tr.T("info_failed"))
		return
	}

	if err := b.store.Save(userID, session.Data{URL: url, Title: info.Title, ID: info.ID}); err != nil {
		b.log.Warn("save session failed", "user", userID, "err", err)
	}

	caption := b.tr.T("choose", info.Title, util.HumanizeDuration(info.DurationSec))
	kb := &models.InlineKeyboardMarkup{InlineKeyboard: [][]models.InlineKeyboardButton{
		{{Text: b.tr.T("btn_audio"), CallbackData: "audio"}},
		{{Text: b.tr.T("btn_video"), CallbackData: "video"}},
	}}
	if _, err := api.SendMessage(ctx, &tgbot.SendMessageParams{ChatID: chatID, Text: caption, ReplyMarkup: kb}); err != nil {
		b.log.Warn("send choices failed", "err", err)
	}
}

func (b *Bot) onCallback(ctx context.Context, api *tgbot.Bot, update *models.Update) {
	q := update.CallbackQuery
	_, _ = api.AnswerCallbackQuery(ctx, &tgbot.AnswerCallbackQueryParams{CallbackQueryID: q.ID})

	choice := q.Data
	if choice != "audio" && choice != "video" {
		return
	}
	if q.Message.Message == nil {
		return // inaccessible message
	}
	userID := q.From.ID
	chatID := q.Message.Message.Chat.ID
	messageID := q.Message.Message.ID

	sess, err := b.store.Load(userID)
	if err != nil || sess == nil {
		b.edit(ctx, api, chatID, messageID, b.tr.T("session_not_found"))
		return
	}

	b.mu.Lock()
	if t := b.tasks[userID]; t != nil && t.active() {
		b.mu.Unlock()
		b.send(ctx, api, chatID, b.tr.T("already_running"))
		return
	}
	if !b.limiter.Allow(userID) {
		b.mu.Unlock()
		mins := int(b.limiter.RetryAfter(userID).Minutes()) + 1
		b.send(ctx, api, chatID, b.tr.T("rate_limited", mins))
		return
	}
	dlCtx, cancel := context.WithCancel(context.Background())
	t := &task{
		status:    "queued",
		choice:    choice,
		title:     sess.Title,
		id:        sess.ID,
		cancel:    cancel,
		createdAt: time.Now(),
		updatedAt: time.Now(),
	}
	b.tasks[userID] = t
	b.mu.Unlock()

	b.edit(ctx, api, chatID, messageID, b.tr.T("queued", sess.Title, choice))
	go b.runDownload(dlCtx, api, userID, chatID, messageID, choice, *sess, t)
}

func (b *Bot) runDownload(ctx context.Context, api *tgbot.Bot, userID, chatID int64, messageID int, choice string, sess session.Data, t *task) {
	defer t.cancel()
	defer func() {
		if b.cfg.CleanupAfterSend {
			download.ForceRemove(t.files)
		}
		_ = b.store.Delete(userID)
	}()

	// Wait for a global download slot (the queue). Cancellation while waiting aborts.
	select {
	case b.slots <- struct{}{}:
		defer func() { <-b.slots }()
	case <-ctx.Done():
		b.setStatus(t, "canceled")
		b.edit(ctx, api, chatID, messageID, b.tr.T("canceled"))
		return
	}

	b.setStatus(t, "downloading")
	b.edit(ctx, api, chatID, messageID, b.tr.T("downloading", sess.Title, choice))

	destDir, err := util.UserDownloadDir(b.cfg.DownloadDir, userID)
	if err != nil {
		b.fail(ctx, api, t, chatID, messageID, err)
		return
	}

	req := download.Request{
		URL:          sess.URL,
		ID:           sess.ID,
		BaseName:     download.BuildBaseName(sess.Title, sess.ID),
		Format:       download.Format(choice),
		MaxHeight:    b.cfg.MaxHeight,
		AudioKbps:    b.cfg.AudioKbitrate,
		MaxMB:        b.cfg.TelegramMaxMB,
		HeightLadder: b.cfg.VideoHeightLadder,
		DestDir:      destDir,
	}

	res, err := b.dl.Fetch(ctx, req, b.progressFn(ctx, api, chatID, messageID))
	if err != nil {
		if ctx.Err() != nil {
			b.setStatus(t, "canceled")
			b.edit(ctx, api, chatID, messageID, b.tr.T("canceled"))
			return
		}
		b.fail(ctx, api, t, chatID, messageID, err)
		return
	}

	b.mu.Lock()
	t.files = res.Files
	t.status = "sending"
	t.updatedAt = time.Now()
	b.mu.Unlock()

	b.sendFiles(ctx, api, chatID, sess.Title, res.Files)
	b.setStatus(t, "done")
	b.edit(ctx, api, chatID, messageID, b.tr.T("completed", sess.Title))
}

// progressFn returns a throttled progress callback that edits the status message.
func (b *Bot) progressFn(ctx context.Context, api *tgbot.Bot, chatID int64, messageID int) download.ProgressFunc {
	var lastPct = -100
	var lastAt time.Time
	return func(p download.Progress) {
		switch p.Stage {
		case download.StageFallbackVid:
			b.send(ctx, api, chatID, b.tr.T("video_too_large", p.Height))
		case download.StageFallbackAud:
			b.send(ctx, api, chatID, b.tr.T("audio_too_large"))
		case download.StageProcessing:
			b.edit(ctx, api, chatID, messageID, b.tr.T("processing", p.Label))
		case download.StageDownloading:
			now := time.Now()
			if p.Percent < lastPct+5 && now.Sub(lastAt) < 2*time.Second {
				return
			}
			lastPct = p.Percent
			lastAt = now
			parts := []string{b.tr.T("progress", p.Label, p.Percent)}
			if p.SpeedBytes > 0 {
				parts = append(parts, util.SizeofFmt(p.SpeedBytes)+"/s")
			}
			if p.ETASeconds > 0 {
				parts = append(parts, fmt.Sprintf("ETA %ds", p.ETASeconds))
			}
			b.edit(ctx, api, chatID, messageID, strings.Join(parts, " • "))
		}
	}
}

func (b *Bot) sendFiles(ctx context.Context, api *tgbot.Bot, chatID int64, title string, files []string) {
	for _, path := range files {
		if err := b.sendOne(ctx, api, chatID, title, path); err != nil {
			b.log.Warn("send file failed", "path", path, "err", err)
			b.send(ctx, api, chatID, b.tr.T("send_failed", filepath.Base(path), err))
		}
	}
}

func (b *Bot) sendOne(ctx context.Context, api *tgbot.Bot, chatID int64, title, path string) error {
	f, err := os.Open(path)
	if err != nil {
		return err
	}
	defer f.Close()
	name := filepath.Base(path)
	upload := &models.InputFileUpload{Filename: name, Data: f}
	lower := strings.ToLower(path)
	switch {
	case strings.HasSuffix(lower, ".mp3"):
		_, err = api.SendAudio(ctx, &tgbot.SendAudioParams{ChatID: chatID, Audio: upload, Title: title})
	case strings.HasSuffix(lower, ".mp4"), strings.HasSuffix(lower, ".mkv"),
		strings.HasSuffix(lower, ".webm"), strings.HasSuffix(lower, ".mov"):
		_, err = api.SendVideo(ctx, &tgbot.SendVideoParams{ChatID: chatID, Video: upload, SupportsStreaming: true})
	default:
		_, err = api.SendDocument(ctx, &tgbot.SendDocumentParams{ChatID: chatID, Document: upload})
	}
	return err
}

func (b *Bot) cmdDownloads(ctx context.Context, api *tgbot.Bot, update *models.Update) {
	if update.Message == nil {
		return
	}
	userID := update.Message.From.ID
	b.mu.Lock()
	t := b.tasks[userID]
	b.mu.Unlock()
	if t == nil {
		b.send(ctx, api, update.Message.Chat.ID, b.tr.T("no_downloads"))
		return
	}
	b.send(ctx, api, update.Message.Chat.ID, b.tr.T("downloads_status",
		t.title, t.choice, t.status,
		t.createdAt.UTC().Format(time.RFC3339), t.updatedAt.UTC().Format(time.RFC3339)))
}

func (b *Bot) cmdCancel(ctx context.Context, api *tgbot.Bot, update *models.Update) {
	if update.Message == nil {
		return
	}
	userID := update.Message.From.ID
	b.mu.Lock()
	t := b.tasks[userID]
	canCancel := t != nil && t.active()
	if canCancel {
		t.cancel()
	}
	b.mu.Unlock()
	if !canCancel {
		b.send(ctx, api, update.Message.Chat.ID, b.tr.T("nothing_to_cancel"))
		return
	}
	b.send(ctx, api, update.Message.Chat.ID, b.tr.T("cancel_requested"))
}

func (b *Bot) cmdClear(ctx context.Context, api *tgbot.Bot, update *models.Update) {
	if update.Message == nil {
		return
	}
	userID := update.Message.From.ID
	chatID := update.Message.Chat.ID
	b.mu.Lock()
	t := b.tasks[userID]
	b.mu.Unlock()
	if t == nil {
		b.send(ctx, api, chatID, b.tr.T("nothing_to_clear"))
		return
	}
	destDir, err := util.UserDownloadDir(b.cfg.DownloadDir, userID)
	removed := 0
	if err == nil && t.id != "" {
		for _, p := range download.FindOutputFiles(destDir, t.id) {
			if os.Remove(p) == nil {
				removed++
			}
		}
	}
	b.mu.Lock()
	t.files = nil
	b.mu.Unlock()
	b.send(ctx, api, chatID, b.tr.T("cleared", removed))
}

// --- helpers ---

func (b *Bot) setStatus(t *task, status string) {
	b.mu.Lock()
	t.status = status
	t.updatedAt = time.Now()
	b.mu.Unlock()
}

func (b *Bot) fail(ctx context.Context, api *tgbot.Bot, t *task, chatID int64, messageID int, err error) {
	b.log.Warn("download failed", "err", err)
	b.setStatus(t, "error")
	b.edit(ctx, api, chatID, messageID, b.tr.T("error", err.Error()))
}

func (b *Bot) send(ctx context.Context, api *tgbot.Bot, chatID int64, text string) {
	if _, err := api.SendMessage(ctx, &tgbot.SendMessageParams{ChatID: chatID, Text: text}); err != nil {
		b.log.Debug("send message failed", "err", err)
	}
}

func (b *Bot) edit(ctx context.Context, api *tgbot.Bot, chatID int64, messageID int, text string) {
	_, err := api.EditMessageText(ctx, &tgbot.EditMessageTextParams{ChatID: chatID, MessageID: messageID, Text: text})
	if err != nil {
		b.send(ctx, api, chatID, text)
	}
}
